src/ui/components/screensaver.c

```diff
@@ -0,0 +1,106 @@
+// Copyright 2019 Shift Cryptosecurity AG
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "screensaver.h"
+
+#include "image.h"
+#include "ui_images.h"
+
+#include <hardfault.h>
+#include <screen.h>
+#include <ui/ui_util.h>
+
+#include <string.h>
+
+static void _render(component_t* component)
+{
+    component_t* image = component->sub_components.sub_components[0];
+
+    static uint16_t counter = 0;
+    counter++;
+
+    // these flip between positive and negative when boncing, can also be used to move multiple
+    // pixels per frame
+    static int8_t x_direction = 1;
+    static int8_t y_direction = 1;
+    // setting relative speed for both axes
+    const int8_t x_slowdown = 1;
+    const int8_t y_slowdown = 1;
+
+    const uint16_t master_slowdown = 6; // for slowing down both x and y in the same amount
+
+    if (counter % (master_slowdown * x_slowdown) == 0) {
```

Comment:
I'd simplify a bit and remove master_slowdown. You have both x_ and y_ nicely defined next to each other. It's very easy to just change both.

Could also be thought of as x_speedup and y_speedup :)
So, maybe a neutral x_speed and y_speed?

Reply:
agree with the simplification, but the naming is accurate: higher value = slower, so `slowdown` is more natural. with `speed`, you would expect higher speed with higher values.

Reply:
done